diff --git a/python/versions.bzl b/python/versions.bzl
index 30929f82..f753df63 100644
--- a/python/versions.bzl
+++ b/python/versions.bzl
@@ -195,6 +195,7 @@ TOOL_VERSIONS = {
             "aarch64-apple-darwin": "d32da9eae3f516cc0bd8240bfef54dede757d6daf1d8cf605eacbc8a205884e8",
             "aarch64-unknown-linux-gnu": "0318b6c9ad6fb229da8d40aa3671ee27eeb678530246a1b172b72071f76091bc",
             "ppc64le-unknown-linux-gnu": "b40b3509dc72abb21f4310f0e94678b36ff73432dc84c41fea132a51c4017f79",
+            "loongarch64-unknown-linux-gnu": "faba8c3e78b1de10bde00d61d2fb9aa57cec509ced869c80c35ea89b95f9d44b",
             "riscv64-unknown-linux-gnu": "a7d847dc62177cf06237dfa26c317148b22418ded51aa89e8cf7242784293ad4",
             "s390x-unknown-linux-gnu": "425abe5d3ec98e9b18c908209a4ffe239a283ee648e0eea65821e45f074689e7",
             "x86_64-apple-darwin": "c1bfab90aea566ffaeff65299a20503a880ea93054bbd8bbed98f4f11e9e7383",
@@ -344,6 +345,7 @@ TOOL_VERSIONS = {
         "sha256": {
             "aarch64-apple-darwin": "a94c02b2d597cd6b075a713fe4e9a909cc97ca6a3b2b2ce86eda21be2062d48e",
             "aarch64-unknown-linux-gnu": "ef7de3b715d519e246d98ff7856247f7f7b357068705f09c6f300b7e7b76c701",
+            "loongarch64-unknown-linux-gnu": "7bb64f1aa937c96b87fa1f4f086958c382c408d833d1c5c11034127ab21a386f",
             "ppc64le-unknown-linux-gnu": "f580efed11cc54e1a221c052e8bc88bfbc12844d3ca8949da828351a1232386e",
             "riscv64-unknown-linux-gnu": "0d7e460e30203a9225b6f417ae972f66415a1cc0e32b37ebc48d195816282669",
             "s390x-unknown-linux-gnu": "d4ada974daadb08a0184c19232ee3b03b3137aa70609760e1a94aaf7b12989ef",
@@ -474,6 +476,7 @@ TOOL_VERSIONS = {
         "sha256": {
             "aarch64-apple-darwin": "d089bfd2c7b98a0942750a195e70d3172beda76d7747097b8afd87028b6e59b6",
             "aarch64-unknown-linux-gnu": "bc57105f8a16acd57b71d926143c7f6ecf61729b40c8b4656f1b98bebd47c710",
+            "loongarch64-unknown-linux-gnu": "a1eb242d96ce0fd5aa1bb9449816a818790b9f8571a476b6f6c6fe69c155ef45",
             "ppc64le-unknown-linux-gnu": "16a0165b0744940702b8fff80b8bf973ac914f78cb6fca28d389583f675e84de",
             "riscv64-unknown-linux-gnu": "d8e62306be8f41c46bcd62ca68f91a1467f47adff632a35ff413dc1043ed56e8",
             "s390x-unknown-linux-gnu": "4e302a4514a73baefdd9b327062bdafeb4115a799deec91c185f6ab45a857241",
@@ -599,6 +602,7 @@ TOOL_VERSIONS = {
             "aarch64-apple-darwin": "8792c4a84c364ab975feca0c27d3157a5435b7baab325a346ae56b223893b661",
             "aarch64-unknown-linux-gnu": "4d7ba5314fab02130d6538f074961ffbf61310cade9180e59026074f9a8939cb",
             "aarch64-pc-windows-msvc": "00bf7d7e8bcf5d1e9c4dfca0247d8e035147777cd57ee9d4c64dedca86b0a464",
+            "loongarch64-unknown-linux-gnu": "f3a8fa7021d6f3e92e8bffa9f6286c4d3ed5debaaff543a14319d8dd59b8e620",
             "ppc64le-unknown-linux-gnu": "2c862eb40a81549d9c11e6bf5a7f07c3406310b14e6a4d16dcdf1c4763ef7090",
             "riscv64-unknown-linux-gnu": "0bb729b95fabd49c7b495f7c44a9086e3970ea57daf66365741574bd36a17e81",
             "s390x-unknown-linux-gnu": "99e465882d217d24ac90e99fac8f32e6a644d0340ac05ee510fb5cdf53f0cfb8",
@@ -770,6 +774,7 @@ TOOL_VERSIONS = {
         "sha256": {
             "aarch64-apple-darwin": "8a1efa6af4e80f08e2c97dda822a3d6c24d6c98e518242f802c6a43ae8401488",
             "aarch64-unknown-linux-gnu": "11fa0591ae2211c08a42ae54944260e36ddf88a1d5604ea0c49e2477be4e5388",
+            "loongarch64-unknown-linux-gnu": "d4a62272459b084c66e5786c01e9abc1c40fd813fac21b272b5c0e8dc6cff090",
             "ppc64le-unknown-linux-gnu": "8dcf34ae1a685fe1893b52917ae04f23328edadc4acae28499d43850c2bdd26c",
             "riscv64-unknown-linux-gnu": "f8ed75aa6cc2011a046be00b629c3c8295267f34280324feaff34c73e7afce39",
             "s390x-unknown-linux-gnu": "7707ee5d19a78bc64ef8a66751ec7f97b64ea06714c7b1b52e8b321c2923ead8",
@@ -781,6 +786,7 @@ TOOL_VERSIONS = {
             "x86_64-unknown-linux-musl": "70076dea0ff65b3c05aae1a97b4a556bf613cc73db30309e59134f9d318f4f7b",
             "aarch64-apple-darwin-freethreaded": "f2143304012e021a603bf1807bf3e4ce163832e43ab9a9829e53cb136497f207",
             "aarch64-unknown-linux-gnu-freethreaded": "d84a7d64c284be387386b9f5da273f6d05486eb6bd8f9e86e2575cb59604cb22",
+            "loongarch64-unknown-linux-gnu-freethreaded": "a425926dbf5109404221da012bf177d5ffacb8a70ea2371a5fc72e8d8486ce68",
             "ppc64le-unknown-linux-gnu-freethreaded": "e76fcaf1bf80a615520dbe7f85ca0bb557fad96d132d836b0ac721e7cc1e2a37",
             "riscv64-unknown-linux-gnu-freethreaded": "24e08a39ba4fc77753e61541e52eed39cc871f4a92a80a3c5dd495056bd8eff9",
             "s390x-unknown-linux-gnu-freethreaded": "1609b223fd38a4a7a4d20e7173d7d9390fe2258f7dd9a15dc9ef0fa49613735d",
@@ -791,6 +797,7 @@ TOOL_VERSIONS = {
         "strip_prefix": {
             "aarch64-apple-darwin": "python",
             "aarch64-unknown-linux-gnu": "python",
+            "loongarch64-unknown-linux-gnu": "python",
             "ppc64le-unknown-linux-gnu": "python",
             "s390x-unknown-linux-gnu": "python",
             "riscv64-unknown-linux-gnu": "python",
@@ -801,6 +808,7 @@ TOOL_VERSIONS = {
             "x86_64-unknown-linux-musl": "python",
             "aarch64-apple-darwin-freethreaded": "python/install",
             "aarch64-unknown-linux-gnu-freethreaded": "python/install",
+            "loongarch64-unknown-linux-gnu-freethreaded": "python/install",
             "ppc64le-unknown-linux-gnu-freethreaded": "python/install",
             "riscv64-unknown-linux-gnu-freethreaded": "python/install",
             "s390x-unknown-linux-gnu-freethreaded": "python/install",
@@ -815,6 +823,7 @@ TOOL_VERSIONS = {
         "sha256": {
             "aarch64-apple-darwin": "016b9eb7c6c41d358a095f52203297812a566376b1e4372571b850f621dc720d",
             "aarch64-unknown-linux-gnu": "bfa5cb2f56032f4ed2c105f5b3b59ea1809672cb74b453e4450399517a594137",
+            "loongarch64-unknown-linux-gnu": "ebcbd61da39a701cba6da1bacfe6b7696f442a2b5a5a6c19fa041951f21f3277",
             "ppc64le-unknown-linux-gnu": "eb8fade967732032be70d5129ed66ad28dfe57e2964550f0f6800dddc1fdcb80",
             "riscv64-unknown-linux-gnu": "82313ee3c45ad0dc825044fef161840dd60277003bf5458da24d46206fff1e09",
             "s390x-unknown-linux-gnu": "5af30600105c42e920e9709afc8deae07c309cd46959c22dc099a1fb45b73902",
@@ -825,6 +834,7 @@ TOOL_VERSIONS = {
             "x86_64-unknown-linux-musl": "af82c7e3ddaebbfb10967c5dd9751bc2fa7f735806a16ebde7600e519c34e587",
             "aarch64-apple-darwin-freethreaded": "d611134bcf090db920d8192ec26e899433cdbae42476fd92ee07cc13b5e32d1f",
             "aarch64-unknown-linux-gnu-freethreaded": "9fcbb8947c07421506187b3f605f34e94c68824d3fd362d84cd1dcdf13285ee3",
+            "loongarch64-unknown-linux-gnu-freethreaded": "747dd9532dc751fe450461a2c1e20cc2b108a7b4363a6284b7fd91551891ef11",
             "ppc64le-unknown-linux-gnu-freethreaded": "8cb6937fb0804ca0d5d867af15b6e7fd05d79d0faa5a6e617e6f6280580a5f66",
             "riscv64-unknown-linux-gnu-freethreaded": "7a235a6f5b814f5ae789fea65b097ad52f840619dfd054135d8ae8443fa8e362",
             "s390x-unknown-linux-gnu-freethreaded": "fee35437df67782b348d57b32cb1acac61384504af28860d552b0d6aeb3ae19e",
@@ -836,6 +846,7 @@ TOOL_VERSIONS = {
         "strip_prefix": {
             "aarch64-apple-darwin": "python",
             "aarch64-unknown-linux-gnu": "python",
+            "loongarch64-unknown-linux-gnu": "python",
             "ppc64le-unknown-linux-gnu": "python",
             "s390x-unknown-linux-gnu": "python",
             "riscv64-unknown-linux-gnu": "python",
@@ -846,6 +857,7 @@ TOOL_VERSIONS = {
             "x86_64-unknown-linux-musl": "python",
             "aarch64-apple-darwin-freethreaded": "python/install",
             "aarch64-unknown-linux-gnu-freethreaded": "python/install",
+            "loongarch64-unknown-linux-gnu-freethreaded": "python/install",
             "ppc64le-unknown-linux-gnu-freethreaded": "python/install",
             "riscv64-unknown-linux-gnu-freethreaded": "python/install",
             "s390x-unknown-linux-gnu-freethreaded": "python/install",
@@ -922,6 +934,17 @@ def _generate_platforms():
             os_name = LINUX_NAME,
             arch = "x86_32",
         ),
+        "loongarch64-unknown-linux-gnu": platform_info(
+            compatible_with = [
+                "@platforms//os:linux",
+                "@platforms//cpu:loongarch64",
+            ],
+            target_settings = [
+                is_libc_glibc,
+            ],
+            os_name = LINUX_NAME,
+            arch = "loongarch64",
+        ),
         "ppc64le-unknown-linux-gnu": platform_info(
             compatible_with = [
                 "@platforms//os:linux",
@@ -1028,6 +1051,9 @@ def get_release_info(platform, python_version, base_url = DEFAULT_RELEASE_BASE_U
         A tuple of (filename, url, archive strip prefix, patches, patch_strip)
     """
 
+    if platform.startswith("loongarch64-unknown-linux-gnu"):
+        base_url = "https://github.com/loong64/python-build-standalone/releases/download"
+
     url = tool_versions[python_version]["url"]
 
     if type(url) == type({}):
@@ -1055,6 +1081,7 @@ def get_release_info(platform, python_version, base_url = DEFAULT_RELEASE_BASE_U
                     "aarch64-pc-windows-msvc": "pgo",
                     "aarch64-unknown-linux-gnu": "lto" if release_id < 20250702 else "pgo+lto",
                     "ppc64le-unknown-linux-gnu": "lto",
+                    "loongarch64-unknown-linux-gnu": "lto",
                     "riscv64-unknown-linux-gnu": "lto",
                     "s390x-unknown-linux-gnu": "lto",
                     "x86_64-apple-darwin": "pgo+lto",

diff --git a/python/private/repo_utils.bzl b/python/private/repo_utils.bzl
--- a/python/private/repo_utils.bzl
+++ b/python/private/repo_utils.bzl
@@ -424,7 +424,9 @@ def _get_platforms_cpu_name(mrctx):
     if arch in ["s390x", "s390"]:
         return "s390x"
     if arch in ["mips64el", "mips64"]:
         return "mips64"
     if arch in ["riscv64"]:
         return "riscv64"
+    if arch in ["loongarch64", "loong64"]:
+        return "loongarch64"
     return arch

diff --git a/python/private/pypi/pip_repository.bzl b/python/private/pypi/pip_repository.bzl
--- a/python/private/pypi/pip_repository.bzl
+++ b/python/private/pypi/pip_repository.bzl
@@ -86,1 +86,2 @@
                 "linux_x86_64",
+                "linux_loongarch64",
diff --git a/python/private/pypi/pep508_env.bzl b/python/private/pypi/pep508_env.bzl
--- a/python/private/pypi/pep508_env.bzl
+++ b/python/private/pypi/pep508_env.bzl
@@ -68,1 +68,2 @@
     "@platforms//cpu:s390x": "s390x",
+    "@platforms//cpu:loongarch64": "loongarch64",

diff --git a/python/private/pypi/whl_installer/platform.py b/python/private/pypi/whl_installer/platform.py
--- a/python/private/pypi/whl_installer/platform.py
+++ b/python/private/pypi/whl_installer/platform.py
@@ -43,2 +43,3 @@ class Arch(Enum):
     ppc64le = 5
     s390x = 6
+    loongarch64 = 8

diff --git a/python/private/pypi/whl_target_platforms.bzl b/python/private/pypi/whl_target_platforms.bzl
--- a/python/private/pypi/whl_target_platforms.bzl
+++ b/python/private/pypi/whl_target_platforms.bzl
@@ -30,2 +30,3 @@
     "ppc64le": "ppc64le",
     "s390x": "s390x",
+    "loongarch64": "loongarch64",
